### PR TITLE
feat: Metal CLI support create shared interconnection with vlans and vrfs

### PR DESCRIPTION
**Pull request by aayushrangwala**
Part of #307

### CODE REVIEW

**Comment by displague** [comment]
Since this will need to be rebased. Perhaps after rebase we can add the 3 or so lines to pick up #388 (keep in mind, this changed with #331).

**Comment by aayushrangwala** [comment]
> Since this will need to be rebased. Perhaps after rebase we can add the 3 or so lines to pick up #388 (keep in mind, this changed with #331).

@displague Yes added the delete force changes in this PR

File: internal/interconnections/create.go [hunk 1]
```diff
@@ -58,14 +84,58 @@ func (c *Client) Create() *cobra.Command {
...
+	if projectID == "" && organizationID == "" {
+		return errors.New("could you provide at least either of projectID OR organizationID")
+	}
+
+	if (vlanFabricVcCreate(vlans) || vrfsFabricVcCreate(vrfs)) && svcTokenType == "" {
+		return errors.New("flag 'service-token-type' is required for vlan or vrfs fabric VC create")
+	}
+
+	if vlanFabricVcCreate(vlans) && vrfsFabricVcCreate(vrfs) {
+		return errors.New("vLans and vrfs both are provided. Please provide any one type of interconnection")
```
Comment (displague) [inline internal/interconnections/create.go]:
```suggestion
		return errors.New("vlans and vrfs both are provided. Please provide any one type of interconnection")
```

File: internal/interconnections/delete.go [hunk 2]
```diff
@@ -3,23 +3,44 @@ package interconnections
...
+  metal interconnections delete -i 7ec86e23-8dcf-48ed-bd9b-c25c20958277
+  >
+  ✔ Are you sure you want to delete device 7ec86e23-8dcf-48ed-bd9b-c25c20958277 [Y/n]: Y
+`,
 
 		RunE: func(cmd *cobra.Command, args []string) error {
 			cmd.SilenceUsage = true
 
+			if !force {
+				fmt.Printf("Are you sure you want to delete device %s [Y/n]: ", connectionID)
```
Comment (displague) [inline internal/interconnections/delete.go]:
```suggestion
				fmt.Printf("Are you sure you want to delete interconnection %s [Y/n]: ", connectionID)
```

or `connection`

File: docs/metal_interconnections_create.md [hunk 3]
```diff
@@ -24,14 +24,17 @@ metal interconnections create -n <name> [-m <metro>] [-r <redundancy> ] [-t <typ
...
+  -h, --help                        help for create
+  -m, --metro string                metro in the interconnection
+  -n, --name string                 Name of the interconnection
+  -O, --organizationID string       Org ID
+  -p, --projectID string            project ID
+  -r, --redundancy string           Website URL of the organization.
+  -T, --service-token-type string   Type of service token for shared connection. Enum: 'a_side', 'z_side'
+      --speed int32                 the maximum speed of the interconnections (default 1000000000)
+  -t, --type string                 type of of interconnection.
+      --vlans int32Slice            Array of int vLANs (default [])
```
Comment (displague) [inline docs/metal_interconnections_create.md]: We generally use `VLANs` (I don't know what's right or best)

File: docs/metal_interconnections_create.md [hunk 4]
```diff
@@ -24,14 +24,17 @@ metal interconnections create -n <name> [-m <metro>] [-r <redundancy> ] [-t <typ
...
-  -p, --projectID string        project ID
-  -r, --redundancy string       Website URL of the organization.
-  -t, --type string             type of of interconnection.
-  -v, --vrfs strings            Return only the specified vrfs.
+  -h, --help                        help for create
+  -m, --metro string                metro in the interconnection
+  -n, --name string                 Name of the interconnection
+  -O, --organizationID string       Org ID
+  -p, --projectID string            project ID
+  -r, --redundancy string           Website URL of the organization.
```
Comment (displague) [inline docs/metal_interconnections_create.md]: This doesn't seem right. (redundacy is a string with a URL as a parameter?)

File: docs/metal_interconnections_create.md [hunk 5]
```diff
@@ -24,14 +24,17 @@ metal interconnections create -n <name> [-m <metro>] [-r <redundancy> ] [-t <typ
...
-  -n, --name string             Name of the interconnection
-  -O, --organizationID string   Org ID
-  -p, --projectID string        project ID
-  -r, --redundancy string       Website URL of the organization.
-  -t, --type string             type of of interconnection.
-  -v, --vrfs strings            Return only the specified vrfs.
+  -h, --help                        help for create
+  -m, --metro string                metro in the interconnection
+  -n, --name string                 Name of the interconnection
+  -O, --organizationID string       Org ID
```
Comment (displague) [inline docs/metal_interconnections_create.md]: Find other examples of organization and project ID arguments and match the description and argument format (`--organization-id`, `--project-id`). "Org ID" is not a good description.

File: docs/metal_interconnections_create.md [hunk 6]
```diff
@@ -24,14 +24,17 @@ metal interconnections create -n <name> [-m <metro>] [-r <redundancy> ] [-t <typ
...
-  -m, --metro string            metro in the interconnection
-  -n, --name string             Name of the interconnection
-  -O, --organizationID string   Org ID
-  -p, --projectID string        project ID
-  -r, --redundancy string       Website URL of the organization.
-  -t, --type string             type of of interconnection.
-  -v, --vrfs strings            Return only the specified vrfs.
+  -h, --help                        help for create
+  -m, --metro string                metro in the interconnection
+  -n, --name string                 Name of the interconnection
```
Comment (displague) [inline docs/metal_interconnections_create.md]: Between Metro and Name, the description format should match. Check the format elsewhere. I think we generally use uppercase and full sentences with punctuation.

File: docs/metal_interconnections_create.md (hunk 3 above)
Comment (displague) [reply docs/metal_interconnections_create.md]: `Array of int VLANs` -> `One or more VLANs to attach to the Interconnection: --vlan 1000 --vlan 1234` (I'm looking at Spot Market options in `metal device create` as insipration)

We'll also want to include an example snippet showing how to define this in the subcommand help.

File: internal/interconnections/create.go [hunk 7]
```diff
@@ -58,14 +84,58 @@ func (c *Client) Create() *cobra.Command {
 	createInterconnectionsCmd.Flags().StringVarP(&metro, "metro", "m", "", "metro in the interconnection")
 	createInterconnectionsCmd.Flags().StringVarP(&redundancy, "redundancy", "r", "", "Website URL of the organization.")
 	createInterconnectionsCmd.Flags().StringVarP(&connType, "type", "t", "", "type of of interconnection.")
-	// createInterconnectionsCmd.Flags().StringVarP(&connType, "serviceTokentype", "T", "", "service token type for interconnection either fabric OR Metal builds")
-	createInterconnectionsCmd.Flags().StringSliceVarP(&vrfs, "vrfs", "v", []string{}, "Return only the specified vrfs.")
+	createInterconnectionsCmd.Flags().StringSliceVar(&vrfs, "vrfs", []string{}, "Array of strings VRF <uuid>.")
 	createInterconnectionsCmd.Flags().StringVarP(&projectID, "projectID", "p", "", "project ID")
 	createInterconnectionsCmd.Flags().StringVarP(&organizationID, "organizationID", "O", "", "Org ID")
+	createInterconnectionsCmd.Flags().Int32SliceVar(&vlans, "vlans", []int32{}, "Array of int vLANs")
```
Comment (displague) [inline internal/interconnections/create.go]: I think the argument name would be singular, and the user specifies multiple using multiple arguments:

`--vlan 1234 --vlan 9876`

I think Cobra also handles `--vlan 1234,9876` for us with `Int32SliceVar`.

I may be wrong about how we've previously named SliceVars. Follow the existing convention, whatever that is.

File: docs/metal_interconnections_create.md (hunk 4 above)
Comment (aayushrangwala) [reply docs/metal_interconnections_create.md]: Yes, that seems incorrect. It seems the already existing flag descriptions for most of them needs to be corrected

File: docs/metal_interconnections_create.md (hunk 3 above)
Comment (aayushrangwala) [reply docs/metal_interconnections_create.md]: Done

File: docs/metal_interconnections_create.md (hunk 6 above)
Comment (aayushrangwala) [reply docs/metal_interconnections_create.md]: Done

File: internal/interconnections/create.go (hunk 7 above)
Comment (aayushrangwala) [reply internal/interconnections/create.go]: Yes right. Making it singulare makes sense

File: test/e2e/ipstest/ips_get_test.go [hunk 8]
```diff
@@ -41,7 +41,7 @@ func TestCli_Ips_Get(t *testing.T) {
 					t.Skip("Skipping this test because someCondition is true")
 				}
 				root := c.Root()
-				projectName := "metal-cli-ips-get" + helper.GenerateRandomString(5)
+				projectName := "metal-cli-ips-get-" + helper.GenerateRandomString(5)
```
Comment (ctreatma) [inline test/e2e/ipstest/ips_get_test.go]: Could you move these changes to a separate PR?

File: internal/interconnections/create.go [hunk 9]
```diff
@@ -2,15 +2,18 @@ package interconnections
...
+	"errors"
 	"fmt"
 
 	metal "github.com/equinix/equinix-sdk-go/services/metalv1"
 	"github.com/spf13/cobra"
 )
 
 func (c *Client) Create() *cobra.Command {
-	var name, metro, redundancy, connType, projectID, organizationID string
+	var name, metro, redundancy, connType, projectID, organizationID, svcTokenType string
```
Comment (ctreatma) [inline internal/interconnections/create.go:13]: We should add e2e or mock tests covering this behavior (this PR may be why you asked about mocking yesterday; I'm just mentioning tests here so it's visible to other reviewers).

**Review by ctreatma — CHANGES_REQUESTED** [review]

File: test/e2e/ipstest/ips_get_test.go (hunk 8 above)
Comment (aayushrangwala) [reply test/e2e/ipstest/ips_get_test.go]: Yeah sure. Removed from this one

File: internal/interconnections/create.go (hunk 9 above)
Comment (aayushrangwala) [reply internal/interconnections/create.go:13]: There were not tests added for any operation interconnections subcommand yet. Not aware about the reason behind it. Adding tests just for new flag would anyway required tests

The mock package I was asking was for packngo to metal-go changes in CPEM as used [currently here](https://github.com/kubernetes-sigs/cloud-provider-equinix-metal/blob/0ac2e0f8959d3f8d7592b6c46651974818f3e242/metal/devices_test.go#L10)

Comment (codinja1188) [reply internal/interconnections/create.go:13]: Initially, the feature is not fully developed, and we plan to submit small PRs for review and testing.

Comment (aayushrangwala) [reply internal/interconnections/create.go:13]: maybe we can cover tests in a separate PR. As anyway the interconnections subcommand is not yet released to prod

Comment (ctreatma) [reply internal/interconnections/create.go:13]: It's fine to introduce tests in a separate PR, but in that case the test PR should be merged first so we can use them to validate this change.

Comment (aayushrangwala) [reply internal/interconnections/create.go:13]: ohkay, I'll work on tests then and park this PR

**Review by codinja1188 — COMMENTED** [review]
Can you add test results?

File: internal/interconnections/create.go (hunk 9 above)
Comment (aayushrangwala) [reply internal/interconnections/create.go:13]: @ctreatma I am thinking, instead of making this PR blocked, we can add tests only to test the new capabilities added in this PR and will create another github issue for tests to be added for older code

Comment (ctreatma) [reply internal/interconnections/create.go:13]: yes, that's a good idea

Comment (aayushrangwala) [reply internal/interconnections/create.go:13]: @ctreatma Due to this bug we discussed yesterday https://github.com/equinix/metal-cli/issues/430
It wont be able to test the create connections from cli as we donot have ID of connection to fetch from server and verify in tests

Should we continue with the PR and wait for the schema level fixes to go in first and then pickup the tests. Anyway these are the new features and not officially released yet

File: test/e2e/interconnections/create_test.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+	outputPkg "github.com/equinix/metal-cli/internal/outputs"
+	"github.com/equinix/metal-cli/test/helper"
+
+	metal "github.com/equinix/equinix-sdk-go/services/metalv1"
+	"github.com/spf13/cobra"
+)
+
+func TestInterconnections_Create(t *testing.T) {
+	subCommand := "interconnections"
+	consumerToken := ""
```
Comment (codinja1188) [inline test/e2e/interconnections/create_test.go]: remove from 21-23
update Line 24
```
rootClient := root.NewClient(helper.ConsumerToken, helper.URL, helper.Version)
```

Comment (aayushrangwala) [reply test/e2e/interconnections/create_test.go]: Done

**Comment by aayushrangwala** [comment]
> Can you add test results?

```
o test -v ./... -timeout 1000s
?   	github.com/equinix/metal-cli/cmd/metal	[no test files]
?   	github.com/equinix/metal-cli/internal/capacity	[no test files]
?   	github.com/equinix/metal-cli/internal/cli	[no test files]
?   	github.com/equinix/metal-cli/internal/completion	[no test files]
?   	github.com/equinix/metal-cli/internal/devices	[no test files]
?   	github.com/equinix/metal-cli/internal/docs	[no test files]
?   	github.com/equinix/metal-cli/internal/emdocs	[no test files]
?   	github.com/equinix/metal-cli/internal/env	[no test files]
?   	github.com/equinix/metal-cli/internal/events	[no test files]
?   	github.com/equinix/metal-cli/internal/facilities	[no test files]
?   	github.com/equinix/metal-cli/internal/gateway	[no test files]
?   	github.com/equinix/metal-cli/internal/hardware	[no test files]
?   	github.com/equinix/metal-cli/internal/init	[no test files]
?   	github.com/equinix/metal-cli/internal/interconnections	[no test files]
?   	github.com/equinix/metal-cli/internal/ips	[no test files]
?   	github.com/equinix/metal-cli/internal/metros	[no test files]
?   	github.com/equinix/metal-cli/internal/organizations	[no test files]
?   	github.com/equinix/metal-cli/internal/os	[no test files]
?   	github.com/equinix/metal-cli/internal/outputs	[no test files]
?   	github.com/equinix/metal-cli/internal/pagination	[no test files]
?   	github.com/equinix/metal-cli/internal/plans	[no test files]
?   	github.com/equinix/metal-cli/internal/ports	[no test files]
?   	github.com/equinix/metal-cli/internal/projects	[no test files]
?   	github.com/equinix/metal-cli/internal/ssh	[no test files]
?   	github.com/equinix/metal-cli/internal/twofa	[no test files]
?   	github.com/equinix/metal-cli/internal/users	[no test files]
?   	github.com/equinix/metal-cli/internal/vlan	[no test files]
?   	github.com/equinix/metal-cli/internal/vrf	[no test files]
=== RUN   TestCli_RegisterCommands
=== RUN   TestCli_RegisterCommands/test
--- PASS: TestCli_RegisterCommands (0.00s)
    --- PASS: TestCli_RegisterCommands/test (0.00s)
PASS
ok  	github.com/equinix/metal-cli/cmd	1.067s
=== RUN   TestCli_Capacity
=== RUN   TestCli_Capacity/get
=== RUN   TestCli_Capacity/get_by_plan_1
=== RUN   TestCli_Capacity/get_by_plan_2
=== RUN   TestCli_Capacity/check_by_multi_metro
=== RUN   TestCli_Capacity/check_by_multi_plan
=== RUN   TestCli_Capacity/check_by_multi_metro_and_plan
--- PASS: TestCli_Capacity (8.49s)
    --- PASS: TestCli_Capacity/get (2.48s)
    --- PASS: TestCli_Capacity/get_by_plan_1 (2.91s)
    --- PASS: TestCli_Capacity/get_by_plan_2 (1.08s)
    --- PASS: TestCli_Capacity/check_by_multi_metro (0.64s)
    --- PASS: TestCli_Capacity/check_by_multi_plan (0.62s)
    --- PASS: TestCli_Capacity/check_by_multi_metro_and_plan (0.76s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/capacitytest	9.361s
?   	github.com/equinix/metal-cli/test/helper	[no test files]
=== RUN   TestCli_Devices_Create_Flags
=== RUN   TestCli_Devices_Create_Flags/create_device
--- PASS: TestCli_Devices_Create_Flags (113.13s)
    --- PASS: TestCli_Devices_Create_Flags/create_device (113.13s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/devices/devicecreateflagstest	114.384s
=== RUN   TestCli_Devices_Create
=== RUN   TestCli_Devices_Create/create_device
--- PASS: TestCli_Devices_Create (156.77s)
    --- PASS: TestCli_Devices_Create/create_device (156.77s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/devices/devicecreatetest	158.444s
=== RUN   TestCli_Devices_Get
=== RUN   TestCli_Devices_Get/get_by_device_id
--- PASS: TestCli_Devices_Get (188.53s)
    --- PASS: TestCli_Devices_Get/get_by_device_id (188.53s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/devices/devicegettest	190.778s
=== RUN   TestCli_Devices_Update
=== RUN   TestCli_Devices_Update/reinstall_device
--- PASS: TestCli_Devices_Update (742.53s)
    --- PASS: TestCli_Devices_Update/reinstall_device (742.53s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/devices/devicereinstalltest	745.504s
=== RUN   TestCli_Devices_Update
=== RUN   TestCli_Devices_Update/start_device
--- PASS: TestCli_Devices_Update (194.28s)
    --- PASS: TestCli_Devices_Update/start_device (194.28s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/devices/devicestarttest	196.901s
=== RUN   TestCli_Devices_Update
=== RUN   TestCli_Devices_Update/stop_device
--- PASS: TestCli_Devices_Update (179.06s)
    --- PASS: TestCli_Devices_Update/stop_device (179.06s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/devices/devicestoptest	182.759s
=== RUN   TestCli_Devices_Update
=== RUN   TestCli_Devices_Update/update_device
--- PASS: TestCli_Devices_Update (181.09s)
    --- PASS: TestCli_Devices_Update/update_device (181.09s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/devices/deviceupdatetest	184.404s
=== RUN   TestCli_Events_Get
=== RUN   TestCli_Events_Get/get_events_by_dev_id
--- PASS: TestCli_Events_Get (189.80s)
    --- PASS: TestCli_Events_Get/get_events_by_dev_id (189.80s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/events/deviceeventstest	193.585s
=== RUN   TestCli_Events_Get
=== RUN   TestCli_Events_Get/get_events_by_proj_id
--- PASS: TestCli_Events_Get (6.50s)
    --- PASS: TestCli_Events_Get/get_events_by_proj_id (6.50s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/events/projecteventstest	10.097s
=== RUN   TestGateways_Create
=== RUN   TestGateways_Create/create_gateways
--- PASS: TestGateways_Create (13.21s)
    --- PASS: TestGateways_Create/create_gateways (2.07s)
=== RUN   TestGateways_Delete
=== RUN   TestGateways_Delete/delete_gateways
--- PASS: TestGateways_Delete (5.00s)
    --- PASS: TestGateways_Delete/delete_gateways (0.95s)
=== RUN   TestGateways_Retrieve
=== RUN   TestGateways_Retrieve/retrieve_gateways_by_projectId
--- PASS: TestGateways_Retrieve (8.19s)
    --- PASS: TestGateways_Retrieve/retrieve_gateways_by_projectId (0.48s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/gateways	27.488s
=== RUN   TestInterconnections_Create
=== RUN   TestInterconnections_Create/create_shared_vlan_interconnection
--- PASS: TestInterconnections_Create (5.72s)
    --- PASS: TestInterconnections_Create/create_shared_vlan_interconnection (3.14s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/interconnections	6.668s
=== RUN   TestCli_Ips_Get
=== RUN   TestCli_Ips_Get/get_ip_reservations
    ips_get_test.go:38: Skipping this test because someCondition is true
--- PASS: TestCli_Ips_Get (0.00s)
    --- SKIP: TestCli_Ips_Get/get_ip_reservations (0.00s)
=== RUN   TestCli_Vlan_Create
=== RUN   TestCli_Vlan_Create/Request_NewIP
    ips_request_test.go:36: Skipping temporarily for now
--- PASS: TestCli_Vlan_Create (0.00s)
    --- SKIP: TestCli_Vlan_Create/Request_NewIP (0.00s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/ipstest	0.938s
=== RUN   TestCli_Metros
=== RUN   TestCli_Metros/get
--- PASS: TestCli_Metros (0.80s)
    --- PASS: TestCli_Metros/get (0.80s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/metrotest	1.807s
=== RUN   TestCli_Organization
=== RUN   TestCli_Organization/get
+--------------------------------------+----------------------------------------------+-------------------------------+
|                  ID                  |                     NAME                     |            CREATED            |
+--------------------------------------+----------------------------------------------+-------------------------------+
| ecd6e156-e2fb-4e5b-b90e-090a055437ee | Equinix Metal Success & Marketing [Official] | 2019-11-15 17:42:22 +0000 UTC |
+--------------------------------------+----------------------------------------------+-------------------------------+
--- PASS: TestCli_Organization (1.05s)
    --- PASS: TestCli_Organization/get (1.05s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/organizationtest	2.529s
=== RUN   TestCli_OperatingSystem
=== RUN   TestCli_OperatingSystem/get
--- PASS: TestCli_OperatingSystem (1.29s)
    --- PASS: TestCli_OperatingSystem/get (1.29s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/ostest	2.898s
=== RUN   TestCli_Plans
=== RUN   TestCli_Plans/get
=== RUN   TestCli_Plans/get_by_slug
=== RUN   TestCli_Plans/get_by_type
--- PASS: TestCli_Plans (3.96s)
    --- PASS: TestCli_Plans/get (2.93s)
    --- PASS: TestCli_Plans/get_by_slug (0.39s)
    --- PASS: TestCli_Plans/get_by_type (0.63s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/plantest	4.911s
=== RUN   TestPorts_Convert
=== RUN   TestPorts_Convert/convert_port_layer-2_bonded_false
=== RUN   TestPorts_Convert/convert_port_layer-2_bonded_true
=== RUN   TestPorts_Convert/convert_port_layer-3_bonded_true
2024/01/15 19:33:38 Converting port f646bac3-1e1e-4a34-9851-0ee6cffd11f7 to layer-3
--- PASS: TestPorts_Convert (233.07s)
    --- PASS: TestPorts_Convert/convert_port_layer-2_bonded_false (23.89s)
    --- PASS: TestPorts_Convert/convert_port_layer-2_bonded_true (9.59s)
    --- PASS: TestPorts_Convert/convert_port_layer-3_bonded_true (10.48s)
=== RUN   TestPorts_Retrieve
=== RUN   TestPorts_Retrieve/retrieve_port
--- PASS: TestPorts_Retrieve (177.83s)
    --- PASS: TestPorts_Retrieve/retrieve_port (0.49s)
=== RUN   TestPorts_VLANs
=== RUN   TestPorts_VLANs/vlan_assignment_port
--- PASS: TestPorts_VLANs (223.54s)
    --- PASS: TestPorts_VLANs/vlan_assignment_port (15.98s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/ports	635.619s
=== RUN   TestCli_Project_Tests
=== RUN   TestCli_Project_Tests/project-create-test
=== RUN   TestCli_Project_Tests/project-update-test
=== RUN   TestCli_Project_Tests/project-get-test
=== RUN   TestCli_Project_Tests/project-get-id-test
=== RUN   TestCli_Project_Tests/project-delete-test
=== RUN   TestCli_Project_Tests/project-bgpenbale-test
=== RUN   TestCli_Project_Tests/project-bgpconfig-test
--- PASS: TestCli_Project_Tests (18.93s)
    --- PASS: TestCli_Project_Tests/project-create-test (3.50s)
    --- PASS: TestCli_Project_Tests/project-update-test (1.89s)
    --- PASS: TestCli_Project_Tests/project-get-test (2.55s)
    --- PASS: TestCli_Project_Tests/project-get-id-test (1.76s)
    --- PASS: TestCli_Project_Tests/project-delete-test (5.79s)
    --- PASS: TestCli_Project_Tests/project-bgpenbale-test (1.49s)
    --- PASS: TestCli_Project_Tests/project-bgpconfig-test (1.95s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/projecttest	23.855s
=== RUN   TestCli_SshKey
=== RUN   TestCli_SshKey/get
--- PASS: TestCli_SshKey (0.71s)
    --- PASS: TestCli_SshKey/get (0.71s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/sshtest	2.026s
=== RUN   TestCli_Users
=== RUN   TestCli_Users/get
--- PASS: TestCli_Users (0.64s)
    --- PASS: TestCli_Users/get (0.64s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/usertest	1.695s
=== RUN   TestCli_Vlan_Create
=== RUN   TestCli_Vlan_Create/create_virtual_network
--- PASS: TestCli_Vlan_Create (1.91s)
    --- PASS: TestCli_Vlan_Create/create_virtual_network (1.91s)
=== RUN   TestCli_Vlan_Clean
=== RUN   TestCli_Vlan_Clean/delete_virtual_network
--- PASS: TestCli_Vlan_Clean (6.81s)
    --- PASS: TestCli_Vlan_Clean/delete_virtual_network (6.81s)
=== RUN   TestCli_Vlan_Get
=== RUN   TestCli_Vlan_Get/get_virtual_network
--- PASS: TestCli_Vlan_Get (2.93s)
    --- PASS: TestCli_Vlan_Get/get_virtual_network (2.93s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/vlan	12.892s
```

File: test/e2e/interconnections/create_test.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+			want: &cobra.Command{},
+			cmdFunc: func(t *testing.T, c *cobra.Command) {
+				root := c.Root()
+				connName := "conn-1-" + randomString
+
+				root.SetArgs([]string{subCommand, "create", "-p", project.GetId(), "--vlan", fmt.Sprintf("%d", vlan.GetVxlan()), "-n", connName, "-m", vlan.GetMetroCode(), "-r", "primary", "-t", "shared", "-T", "a_side", "-s", "50000000"})
+
+				out := helper.ExecuteAndCaptureOutput(t, root)
+
+				conns, err := pager.GetAllProjectInterconnections(*apiClient.InterconnectionsApi, project.GetId(), []string{}, []string{})
```
Comment (ctreatma) [inline test/e2e/interconnections/create_test.go]: Two notes here:

1. Based on what the portal is doing, we could instead get organization interconnections and specify the project ID as a request parameter in order to get only the interconnections that belong to the test project.
2. The `pager` functions should not be used any more; as of now, the SDK should be providing its own paginator functions (`ExecuteWithPagination` instead of `Execute`), and we should use those instead of making/maintaining our own in the CLI

Comment (aayushrangwala) [reply test/e2e/interconnections/create_test.go]: I tried using to get the connections for whole org but for project, when created using `default org` organization object is not populated in its model. Even after fetching it explicitly.

Also, for fetching all connections for an org, there is no pagination it seems as there is no `ExecuteWithPagination` nor `pager.GetAllOrganizationInterconnections()`

Added the fix to fetch default org of current user and use that ID to get conns for an org

Comment (aayushrangwala) [reply test/e2e/interconnections/create_test.go]: @ctreatma I have used `org connections` in the tests but still failing with the [same schema error](https://github.com/equinix/metal-cli/actions/runs/7558522078/job/20581385766?pr=397#step:5:111). The problem here is if there is any connection for a project or Org which was created with speed out of `int32` limit, is failing to get via metal-go.

My local tests were passing because I deleted all the other interconnection with more speeds.

Comment (ctreatma) [reply test/e2e/interconnections/create_test.go]: There's a draft PR up to work around this API issue within the SDK; could you try it out and confirm if it resolves your issue? https://github.com/equinix/equinix-sdk-go/pull/22

Comment (ctreatma) [reply test/e2e/interconnections/create_test.go]: The interconnection create test is failing but I suspect it's related to an ongoing incident because the error is a generic HTTP 500 error.  We'll have to wait until the incident is resolved and then re-run.  Once the patch is validated it can be merged and released, and this PR can be updated to use the latest SDK release.

File: test/e2e/interconnections/create_test.go [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+				var conn *metal.Interconnection
+				for index, c := range conns.GetInterconnections() {
+					if c.GetName() == connName {
+						conn = &conns.GetInterconnections()[index]
+						break
+					}
+				}
+
+				t.Cleanup(func() {
+					helper.CleanupInterconnection(t, conn.GetId())
```
Comment (ctreatma) [inline test/e2e/interconnections/create_test.go:65]: Test is failing on this line: https://github.com/equinix/metal-cli/actions/runs/7574424576/job/20634890241#step:5:115

```
=== RUN   TestInterconnections_Create/create_shared_vlan_interconnection
    create_test.go:75: Error when calling `InterconnectionsApi.DeleteVirtualCircuit`` for cbf4bea9-9442-4ac6-a0d7-35c5d0d4e8b4: 422 Unprocessable Entity
--- FAIL: TestInterconnections_Create (3.20s)
    --- FAIL: TestInterconnections_Create/create_shared_vlan_interconnection (1.97s)
FAIL
```

I _think_ this is expected, because it looks like the helper will try to delete the virtual circuits for the interconnection, but the virtual circuits on a shared interconnection are managed by Metal and cannot be deleted by the user.

File: test/e2e/interconnections/create_test.go (hunk 11 above)
Comment (ctreatma) [reply test/e2e/interconnections/create_test.go]: When I re-ran the test it still failed, but this time with a 422 error during cleanup, which indicates that the cleanup code is doing something it shouldn't do.  I've released v0.31.1 of equinix-sdk-go, so you can pull that in to your PR and that should resolve the `int32` errors, but you will still need to dig into the test cleanup failure.

Comment (aayushrangwala) [reply test/e2e/interconnections/create_test.go]: Yes, I'll check the cleanup failure issue

**Comment by aayushrangwala** [comment]
```PASS
coverage: 9.8% of statements in ./...
ok  	github.com/equinix/metal-cli/test/e2e/interconnections	4.865s	coverage: 9.8% of statements in ./...
```

Tests are passing for the changes in this PR, hence we can merge it

**Review by ctreatma — APPROVED** [review]
The only test failure I see in CI is a platform issue affecting a pre-existing device test that is not modified in this PR.
